src/webgpu/shader/execution/expression/call/builtin/atomics/atomicCompareExchangeWeak.spec.ts

```diff
@@ -34,6 +37,273 @@ struct __atomic_compare_exchange_result<T> {
 `
   )
   .params(u =>
-    u.combine('SC', ['storage', 'uniform'] as const).combine('T', ['i32', 'u32'] as const)
+    u
+      .combine('workgroupSize', workgroupSizes)
+      .combine('dispatchSize', dispatchSizes)
+      .combine('mapId', keysOf(kMapId))
+      .combine('scalarType', ['u32', 'i32'])
   )
-  .unimplemented();
+  .fn(async t => {
+    const numInvocations = t.params.workgroupSize * t.params.dispatchSize;
+    const bufferNumElements = numInvocations;
+    const scalarType = t.params.scalarType;
+    const mapId = kMapId[t.params.mapId];
+    const extra = mapId.wgsl(numInvocations, t.params.scalarType); // Defines map_id()
+
+    const wgsl =
+      `
+      @group(0) @binding(0)
+      var<storage, read_write> input : array<atomic<${scalarType}>>;
+
+      @group(0) @binding(1)
+      var<storage, read_write> output : array<${scalarType}>;
+
+      @group(0) @binding(2)
+      var<storage, read_write> exchanged : array<${scalarType}>;
+
+      @compute @workgroup_size(${t.params.workgroupSize})
+      fn main(
+          @builtin(global_invocation_id) global_invocation_id : vec3<u32>,
+          ) {
+        let id = ${scalarType}(global_invocation_id[0]);
+
+        // Exchange every third value
+        var comp = id + 1;
+        if (id % 3 == 0) {
+          comp = id;
+        }
+        let r = atomicCompareExchangeWeak(&input[id], comp, map_id(id * 2));
```

Comment:
While this is a good basic test for checking the comparison behaviour - the one-invocation to one-memory-location aspect is not really exercising the atomic RMW properties of this builtin.

For example, you could replace this line with:

```ts

        let r = atomicLoad(&wg[id]);
        let eq = r == comp;
        if (eq) {
          atomicStore(&wg[id], map_id(global_id * 2));
        }
```

And with minor tweaks to the usage of `r`, the test will still pass.

I recommend adding another test for multiple-invocations to one-memory-location.